lightning/src/chain/channelmonitor.rs

```diff
@@ -2001,7 +1801,41 @@ impl<ChanSigner: ChannelKeys> ChannelMonitor<ChanSigner> {
 
 		self.onchain_tx_handler.block_disconnected(height, broadcaster, fee_estimator, logger);
 
-		self.last_block_hash = block_hash.clone();
+		self.last_block_hash = block_hash;
+	}
+
+	/// Filters a block's `txdata` for transactions spending watched outputs or for any child
+	/// transactions thereof.
+	fn filter_block<'a>(&self, txdata: &TransactionData<'a>) -> Vec<&'a Transaction> {
+		let mut matched_txn = HashSet::new();
+		txdata.iter().filter(|&&(_, tx)| {
+			let mut matches = self.spends_watched_output(tx);
+			for input in tx.input.iter() {
+				if matches { break; }
+				if matched_txn.contains(&input.previous_output.txid) {
+					matches = true;
+				}
+			}
+			if matches {
+				matched_txn.insert(tx.txid());
+			}
+			matches
+		}).map(|(_, tx)| *tx).collect()
+	}
+
+	/// Checks if a given transaction spends any watched outputs.
+	fn spends_watched_output(&self, tx: &Transaction) -> bool {
+		for input in tx.input.iter() {
+			for (txid, outputs) in self.get_outputs_to_watch().iter() {
```

Comment:
This is a HashMap, we don't have to iterate over it :).

Reply:
Right.. too much literal translation on my part. :P